pkg/utils/certificateUtils.go

```diff
@@ -17,33 +17,97 @@ package utils
 
 import (
 	"crypto/x509"
+	"io/fs"
 	"os"
+	"strings"
 
 	"path/filepath"
 
+	"github.com/deislabs/ratify/pkg/homedir"
 	"github.com/notaryproject/notation-go-lib/crypto/cryptoutil"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
 )
 
 func GetCertificatesFromPath(path string) ([]*x509.Certificate, error) {
-
 	var certs []*x509.Certificate
+	fileMap := map[string]bool{} //a map to track path of physical files
+
+	path = ReplaceHomeShortcut(path)
 
 	err := filepath.Walk(path, func(file string, info os.FileInfo, err error) error {
-		if !info.IsDir() {
-			cert, certError := cryptoutil.ReadCertificateFile(file) // ReadCertificateFile returns empty if file was not a certificate
-			if certError != nil {
-				return certError
+
+		physicalFileInfo := info
+		physicalFilePath := file
+
+		if info == nil || err != nil {
+			logrus.Warnf("Invalid path '%v' skipped, error %v", file, err)
+			return nil
+		}
+
+		// In a cluster environment, each mounted file results in a physical file and a symlink
+		// check if file is a link and get the actual file path
+		if info.Mode()&os.ModeSymlink != 0 {
+
+			physicalFilePath, err = filepath.EvalSymlinks(file)
+
+			if err != nil {
+				logrus.Errorf("error evaluating symbolic link %v , error '%v'", file, file)
+				return nil
+			}
+
+			physicalFileInfo, err = os.Lstat(physicalFilePath)
+
+			if err != nil {
+				logrus.Errorf("error getting file info for path '%v', error '%v'", physicalFilePath, err)
+				return nil
 			}
-			if cert != nil {
-				certs = append(certs, cert...)
+		}
+
+		if physicalFileInfo != nil && !physicalFileInfo.IsDir() {
+			if _, ok := fileMap[physicalFilePath]; !ok {
+				err = loadCertFile(physicalFileInfo, physicalFilePath, &certs, fileMap)
+				if err != nil {
+					return errors.Wrap(err, "error reading certificate file "+physicalFilePath)
+				}
 			}
 		}
+
 		return nil
 	})
 
 	if err != nil {
```

Comment:
what is this err for ? when all errs were local and were handled?

Reply:
this is the err returned from walk right? If so, please ignore this comment

Reply:
Correct , the walk it self might throw errors.